### ISSUE TITLE
ProofGeneral affects line height with some fonts

### ISSUE THREAD

**Issue by RalfJung** [ProofGeneral/PG#150]
For some strange reason, it seems like opening the same file with ProofGeneral and with "plain" emacs results in differences in the line height -- in particular, with ProofGeneral, not all lines have the same height, which is somewhat irritating.

Consider the following text:
```
  Context `{Cofe A, Cofe B, !Inhabited A, !Inhabited B}.
  Context (fA fA' : A → B → A).
  Context (fB fB' : A → B → B).
  Context `{∀ n, Proper (dist_later n ==> dist n ==> dist n) fA}.
  Context `{∀ n, Proper (dist_later n ==> dist n ==> dist n) fA'}.
  Context `{∀ n, Proper (dist_later n ==> dist_later n ==> dist n) fB}.
  Context `{∀ n, Proper (dist_later n ==> dist_later n ==> dist n) fB'}.
```
When put into a file bla.txt, emacs shows this as follows:
![spectacle n10079](https://cloud.githubusercontent.com/assets/330628/22291171/309af74e-e305-11e6-9f88-e0cd1c59df1c.png)
When I call the file bla.v, it renders as follows:
![spectacle c10079](https://cloud.githubusercontent.com/assets/330628/22291183/3c35834e-e305-11e6-92a1-7f9c2942796c.png)
As you can see, the lines with a universal quantifier are higher now.

This is with [Fira Code](https://github.com/tonsky/FiraCode) as the font, but Fira Mono shows the same problem. I am not sure whether the font is to blame (but no other editor does this), or emacs (but it can render things fine in other buffers), or PG (but PG doesn't touch font rendering, does it...?).

Also, this started somewhat recently. Not sure what changed to make this happen, I only know the font did not change.

**Comment by RalfJung** [comment]
I just learned that emacs actually picks the font per-character, and that choice seems to be altered by something when I am in a .v file. In a .txt file, the universal quantifier is `-unknown-DejaVu Math TeX Gyre-normal-normal-normal-*-13-*-*-*-*-0-iso10646-1`, but in a .v file, it is `-unknown-STIX-bold-normal-normal-*-13-*-*-*-*-0-iso10646-1`. That would explain the difference. Why it doesn't use the font I configure (Fira Code) everywhere, I cannot tell.

**Comment by RalfJung** [comment]
I tried uninstall company-coq, and even everything called company*, to no avail.

**Comment by RalfJung** [comment]
Possibly related: https://lists.gnu.org/archive/html/bug-gnu-emacs/2015-05/msg00696.html

**Comment by cpitclaudel** [comment]
Font selection is a bit unpredictable.  If you're using literal `∀` symbols in your source code, then the issue is probably due to the fact that `coq-mode` displays them in bold.  The font that Emacs uses in `.txt` probably doesn't have a bold variant, so Emacs falls back to a different one which does.

I don't think you're seeing an instance of the bug you linked to (which has been fixed in Emacs 25).

> Why it doesn't use the font I configure (Fira Code) everywhere, I cannot tell.

Does Fira Code include `∀`? If not, that's the reason :)

There are tips on company-coq's webpage about setting up font to avoid this kind of surprises.  Alternatively, you might be interested in https://github.com/cpitclaudel/monospacifier

In any case, I don't think this is a PG bug :)

**Comment by RalfJung** [comment]
> Does Fira Code include ∀? If not, that's the reason :)

Indeed it does not.

> In any case, I don't think this is a PG bug :)

I guess you are right. Closing this.

For the record, my solution was to remove the `custom-set-font` from my `.emacs` file and instead add the following:
```
(let ((font-sets '("fontset-default"
                   "fontset-standard"
                   "fontset-startup")))
  (mapcar
   (lambda (font-set)
     ;; all the characters in that range (which is the full possible range)
     (set-fontset-font font-set '(#x000000 . #x3FFFFF) (font-spec :family "Monospace" :weight 'medium :size 10.0))
     ;; for all missing characters, fall back
     (set-fontset-font font-set nil "DejaVu Sans Mono"))
   font-sets))
```
This seems to be the must succinct way to tell emacs to use the "Monospace" font, and if there's a symbol missing there, fall back to "DejaVu Sans Mono". That fixes the trouble reported here, and also lots of other problems with characters not having uniform width.

**Comment by cpitclaudel** [comment]
> EDIT: Thanks to your snippets from the monospacifier, I was able to shorten this to […]

:)

**Comment by RalfJung** [comment]
Actually, I had to further change this to
```
(dolist (ft (fontset-list))
  (set-fontset-font ft 'unicode (font-spec :name "Monospace" :weight 'medium :size 10.0))
  (set-fontset-font ft nil (font-spec :name "DejaVu Sans Mono"))
  )
```
The snippet I used previously caused emacs to slow down horribly, to the extend that after doing "C-c C-b", files would take something like 2 to 5 times as much as usual to process. Even navigating some files got slow. Almost as if emacs spent ridiculous amounts of CPU time on deciding which characters to render how.
Somehow, not using `'append` fixes that. Whatever...
